builtins/target-avx512skx-x32.ll

```diff
@@ -56,7 +56,100 @@ define i1 @__none(<WIDTH x MASK> %mask) nounwind readnone alwaysinline {
+    %perm_high = shufflevector <32 x i32> %perm, <32 x i32> undef, <16 x i32> <i32 16, i32 17, i32 18, i32 19, i32 20, i32 21, i32 22, i32 23, i32 24, i32 25, i32 26, i32 27, i32 28, i32 29, i32 30, i32 31>
+
+    ; Two 512-bit VPERMI2D operations
+    %result1 = call <16 x i32> @llvm.x86.avx512.vpermi2var.d.512(<16 x i32> %low, <16 x i32> %perm_low, <16 x i32> %high)
+    %result2 = call <16 x i32> @llvm.x86.avx512.vpermi2var.d.512(<16 x i32> %low, <16 x i32> %perm_high, <16 x i32> %high)
+
+    ; Concatenate results
+    %final = shufflevector <16 x i32> %result1, <16 x i32> %result2, <32 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15, i32 16, i32 17, i32 18, i32 19, i32 20, i32 21, i32 22, i32 23, i32 24, i32 25, i32 26, i32 27, i32 28, i32 29, i32 30, i32 31>
+    ret <32 x i32> %final
+}
+
+declare <16 x float> @llvm.x86.avx512.vpermi2var.ps.512(<16 x float> %a, <16 x i32> %idx, <16 x float> %b)
+define <32 x float> @__shuffle_float(<32 x float> %input, <32 x i32> %perm) {
+    ; Split input into two 512-bit halves (16 x float each)
+    %low = shufflevector <32 x float> %input, <32 x float> undef, <16 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15>
+    %high = shufflevector <32 x float> %input, <32 x float> undef, <16 x i32> <i32 16, i32 17, i32 18, i32 19, i32 20, i32 21, i32 22, i32 23, i32 24, i32 25, i32 26, i32 27, i32 28, i32 29, i32 30, i32 31>
+
+    ; Get first 16 and second 16 indices
+    %perm_low = shufflevector <32 x i32> %perm, <32 x i32> undef, <16 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15>
+    %perm_high = shufflevector <32 x i32> %perm, <32 x i32> undef, <16 x i32> <i32 16, i32 17, i32 18, i32 19, i32 20, i32 21, i32 22, i32 23, i32 24, i32 25, i32 26, i32 27, i32 28, i32 29, i32 30, i32 31>
+
+    ; Two 512-bit VPERMI2PS operations
+    %result1 = call <16 x float> @llvm.x86.avx512.vpermi2var.ps.512(<16 x float> %low, <16 x i32> %perm_low, <16 x float> %high)
+    %result2 = call <16 x float> @llvm.x86.avx512.vpermi2var.ps.512(<16 x float> %low, <16 x i32> %perm_high, <16 x float> %high)
+
+    ; Concatenate results
+    %final = shufflevector <16 x float> %result1, <16 x float> %result2, <32 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7, i32 8, i32 9, i32 10, i32 11, i32 12, i32 13, i32 14, i32 15, i32 16, i32 17, i32 18, i32 19, i32 20, i32 21, i32 22, i32 23, i32 24, i32 25, i32 26, i32 27, i32 28, i32 29, i32 30, i32 31>
+    ret <32 x float> %final
+}
+
+declare <32 x i64> @llvm.masked.gather.v32i64.v32p0i64(<32 x i64*>, i32, <32 x i1>, <32 x i64>)
+define <32 x i64> @__shuffle_i64(<32 x i64> %input, <32 x i32> %perm) nounwind readnone alwaysinline {
+  ; Store input vector to memory so we can gather from it
+  %input_alloca = alloca [32 x i64], align 64
```

Comment:
Big allocation with alignment can be slow. Have you compared this implementation with multiple 4x shufflevectors?

Reply:
> Big allocation with alignment can be slow. 

This is stack memory. 

> Have you compared this implementation with multiple 4x shufflevectors?

I'm not quite sure which implementation with 4x shufflevectors you are referring to.

Reply:
I mean something like this:
```
 %input_0 = shufflevector <32 x i64> %input, <32 x i64> undef, <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
  %input_1 = ...
  %input_2 = ...
  %input_3 = ...

;   Same for permutation indexes
  %perm_0 = shufflevector <32 x i32> %perm, <32 x i32> undef, <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
  %perm_1 = ...
  %perm_2 = ...
  %perm_3 = ...

  ; Perform 4 separate 8-wide shuffles using vpermi2q 
  ; Combine results
```